sbnana/CAFAna/Systs/EnergySysts.cxx

```diff
@@ -0,0 +1,211 @@
+#include "sbnana/CAFAna/Systs/EnergySysts.h"
+#include "sbnanaobj/StandardRecord/Proxy/SRProxy.h"
+
+namespace ana {
+
+  void EnergyScaleSyst::Shift(double sigma, caf::SRSliceProxy *sr, double& weight) const
+  {
+    double scale = uncertainty * sigma;
+    auto& det = sr->truth.det;
+    bool all = (detector == EnergyScaleSystDetector::kAll);
+    bool nd = (detector == EnergyScaleSystDetector::kSBND && det == caf::kSBND);
+    bool ub = (detector == EnergyScaleSystDetector::kMicroBooNE && det == caf::kUNKNOWN);
+    bool fd = (detector == EnergyScaleSystDetector::kICARUS && det == caf::kICARUS);
+    bool detector_cut = all || nd || ub || fd;
+    if(!sr->truth.iscc || abs(sr->truth.pdg) != 14 || isnan(sr->fake_reco.nuE) || !detector_cut) 
+      return ;
+    double particle_energy = 0.0;
+    switch(part) {
+    case EnergyScaleSystParticle::kMuon:
+      particle_energy += sr->fake_reco.lepton.ke;
+      break;
+    case EnergyScaleSystParticle::kHadron:
+      for(const auto& hadron: sr->fake_reco.hadrons) {
+        particle_energy += hadron.ke;
+      }
+      break;
+    case EnergyScaleSystParticle::kNeutron:
+      for(const auto& hadron: sr->fake_reco.hadrons) {
+        if(hadron.pid == 2112) {
+          particle_energy += hadron.ke;
+        }
+      }
+      break;
+    case EnergyScaleSystParticle::kEM:
+      for(const auto& hadron: sr->fake_reco.hadrons) {
+        if(hadron.pid == 111) {
+          particle_energy += hadron.ke;
+        }
+      }
+      break;
+   case EnergyScaleSystParticle::kChargedHadron:
+      for(const auto& hadron: sr->fake_reco.hadrons) {
+        auto pid = hadron.pid;
+        if(pid == 2212 || abs(pid) == 211) {
+          particle_energy += hadron.ke;
+        }
+      }
+      break;
+    }
+    switch(term) {
+    case EnergyScaleSystTerm::kConstant:
+      break;
+    case EnergyScaleSystTerm::kSqrt:
+      scale *= std::sqrt(particle_energy);
+      break;
+    case EnergyScaleSystTerm::kInverseSqrt:
+      scale /= std::sqrt(particle_energy + 0.1);
+      break;
+    }
+    sr->fake_reco.nuE += particle_energy * scale;
+   
+  }
+
+   const EnergyScaleSyst kEnergyScaleMuon(EnergyScaleSystTerm::kConstant, EnergyScaleSystParticle::kMuon, EnergyScaleSystDetector::kAll, 0.02, "EnergyScaleMuon", "Correlated 2% scale systematic on E_{#mu}");
+   const EnergyScaleSyst kEnergyScaleMuonSqrt(EnergyScaleSystTerm::kSqrt, EnergyScaleSystParticle::kMuon, EnergyScaleSystDetector::kAll, 0.02, "EnergyScaleMuonSqrt", "Correlated 2% scale systematic on E_{#mu} * #sqrt{E_{#mu}}");
```

Comment:
Can use `#times` to get a proper multiplication symbol.

Reply:
Are these times and divides mathematically correct descriptions of the uncertainty? Otherwise maybe just put linear/sqrt/inv sqrt in the names.

Reply:
The code does `E += E*scale*sqrt(E)` and I think the names I had accurately reflected that, but I changed them anyway to avoid any confusion.

Reply:
The name implied `E*sqrt(E) *= 1+scale;` to me, which I don't think is the same thing.